sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala

```diff
@@ -95,6 +95,11 @@ object ReplaceNullWithFalseInPredicate extends Rule[LogicalPlan] {
   private def replaceNullWithFalse(e: Expression): Expression = e match {
     case Literal(null, BooleanType) =>
       FalseLiteral
+    case Not(In(value, list)) if (value +: list).exists(isNullLiteral) =>
```

Comment:
`null in (xxx)`  is a special case that we have already handled in `NullPropagation`, but it's better that return the same result whether other rule is excluded. We also add the `value` pattern here.

Reply:
Can we add some comments to explain the rationale?

if `value` is null, the IN result is null and thus `Not(null)` means false. If `value` is not null but the list contains null values, the IN result can either be `true` or `null`, and thus the final result after Not is always false as a predicate.

Reply:
Yes, it's better to add the comment.